### ISSUE TITLE
The Request property of ODataMediaTypeFormatter in HttpConfiguration.Formatters get modified.

### ISSUE THREAD

**Issue by kfstorm** [rbeauchamp/Swashbuckle.OData#93]
I recently found that a new We API action I wrote returns 500 error when deployed, but it's totally OK in my unit test. After digging for a whole day, I found that the 500 error only get triggered if I visited swagger doc page before. A fresh new website instance can return 200 until I visit the swagger doc URL. Finally I found that it's because the swagger doc generation code modified the `Request` property of `ODataMediaTypeFormatter` in `HttpConfiguration.Formatters`.

We have a custom `ODataSerializerProvider` to deal with empty query result of `SingleResult`. See http://odata.github.io/WebApi/#10-02-work-around-for-singleresult-create. And this bug can be triggered only if we use this customized `ODataSerializerProvider` because it returned non-null serializer for some scenarios.

The bug is triggered as below steps:

1. Visiting swagger doc URL causes the `Request` property of `ODataMediaTypeFormatter` changed to the last request generated by Swashbuckle.OData. E.g. http://any/odata/Products. The code is [here](https://github.com/rbeauchamp/Swashbuckle.OData/blob/master/Swashbuckle.OData/Descriptions/ODataActionDescriptorMapperBase.cs#L97).

2. When client visits the Web API controller, and We API controller returns a result, ASP.NET will find the best match formatter to serialize the result.

    1. Firstly in `NegotiatedContentResult.Execute` method, ASP.NET will call formatter's `CanWriteType(Type type)` method. In the implementation of `ODataMediaTypeFormatter`, it reads `IEdmModel` and only if the model is not empty, it goes into `ODataSerializerProvider.GetODataPayloadSerializer`. Because Swashbuckle.OData changed the Request of formatter to http://any/odata/Products, the `IEdmModel` instance can be read from the request and not null.
    2. Then in the custom implementation of `ODataSerializerProvider`, `NullSerializerProvider`, it returns a custom serializer instead of null serializer returned by base provider. So here it returns the custom serializer for our Web API response data type.
    3. Finally ASP.NET do the real serialization work by creating a new instance of `ObjectContent` with a new formatter created by `MediaTypeForamtter.GetPerRequestFormatterInstance`. Here is the problem. In the constructor of `ObjectContent`, it checks `MediaTypeFormatter.CanWriteType` again. Because the formatter is a another copy of original with executing request set to the `Request` property, and the request is not an OData request but a Web API request, the `IEdmModel` instance doens't exist in the request. So `MediaTypeFormatter.CanWriteType` returns false this time. `ObjectContent` hence throws exception:

> System.InvalidOperationException: The configured formatter 'System.Web.OData.Formatter.ODataMediaTypeFormatter' cannot write an object of type 'Foobar'.

I know it's a bad idea to return a custom serializer for all types. I can restrict the custom serializer to `SingleResult` type only. But I also think it's not right to change the `Request` property of `ODataMediaTypeFormatter` instances stored in `HttpConfiguration.Formatters`. We expect `ODataMediaTypeFormatter` only processes OData requests. But Swashbuckle.OData breaks our expectation.

**Comment by kfstorm** [comment]
Here is my modification of unit test to repro the bug.
```
diff --git a/Swashbuckle.OData.Tests/Fixtures/ModelSharedBetweenODataAndWebApiTests.cs b/Swashbuckle.OData.Tests/Fixtures/ModelSharedBetweenODataAndWebApiTests.cs
index 8c2feea..a161295 100644
--- a/Swashbuckle.OData.Tests/Fixtures/ModelSharedBetweenODataAndWebApiTests.cs
+++ b/Swashbuckle.OData.Tests/Fixtures/ModelSharedBetweenODataAndWebApiTests.cs
@@ -2,14 +2,21 @@
 using System.Collections.Generic;
 using System.ComponentModel.DataAnnotations;
 using System.Linq;
+using System.Net;
+using System.Net.Http;
 using System.Threading.Tasks;
 using System.Web.Http;
 using System.Web.Http.Description;
 using System.Web.OData;
 using System.Web.OData.Builder;
 using System.Web.OData.Extensions;
+using System.Web.OData.Formatter;
+using System.Web.OData.Formatter.Deserialization;
+using System.Web.OData.Formatter.Serialization;
 using FluentAssertions;
+using Microsoft.OData.Core;
 using Microsoft.OData.Edm;
+using Microsoft.Owin;
 using Microsoft.Owin.Hosting;
 using NUnit.Framework;
 using Owin;
@@ -43,6 +50,23 @@ namespace Swashbuckle.OData.Tests
                 await ValidationUtils.ValidateSwaggerJson();
             }
         }
+
+        [Test]
+        public async Task TestCustomApiModels()
+        {
+            Action<SwaggerDocsConfig> config = c => c.DocumentFilter<ApplySharedModelsDocumentation>();
+            using (WebApp.Start(HttpClientUtils.BaseAddress, appBuilder => SharedModelsSetup.Configuration(appBuilder, config, typeof(SharedModelsSetup.SharedModelsController), typeof(SharedModelsSetup.SharedModelsWebApiController))))
+            {
+                // If remove this line, then the test case can pass.
+                await ValidationUtils.ValidateSwaggerJson();
+
+                // Arrange
+                var httpClient = HttpClientUtils.GetHttpClient(HttpClientUtils.BaseAddress);
+                // Verify that the WebApi route in the test controller is valid
+                var response = await httpClient.GetAsync("CustomApiModels");
+                Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+            }
+        }
     }
 
     public class SharedModelsSetup
@@ -58,6 +82,9 @@ namespace Swashbuckle.OData.Tests
 
             config = ConfigureOData(appBuilder, targetControllers, config, unitTestConfigs);
 
+            config.Formatters.InsertRange(0, ODataMediaTypeFormatters.Create(new NullSerializerProvider(), new DefaultODataDeserializerProvider()));
+            //config.Formatters.InsertRange(0, ODataMediaTypeFormatters.Create());
+
             config.EnsureInitialized();
         }
 
@@ -86,6 +113,64 @@ namespace Swashbuckle.OData.Tests
             return builder.GetEdmModel();
         }
 
+        public class NullSerializerProvider : DefaultODataSerializerProvider
+        {
+            private readonly NullEntityTypeSerializer _nullEntityTypeSerializer;
+
+            public NullSerializerProvider()
+            {
+                _nullEntityTypeSerializer = new NullEntityTypeSerializer(this);
+            }
+
+            public override ODataSerializer GetODataPayloadSerializer(IEdmModel model, Type type, HttpRequestMessage request)
+            {
+                var serializer = base.GetODataPayloadSerializer(model, type, request);
+                if (serializer == null)
+                {
+                    var functions = model.SchemaElements.Where(s => s.SchemaElementKind == EdmSchemaElementKind.Function
+                                                                    || s.SchemaElementKind == EdmSchemaElementKind.Action);
+                    var isFunctionCall = false;
+                    foreach (var f in functions)
+                    {
+                        // ReSharper disable once UseStringInterpolation
+                        var fname = string.Format("{0}.{1}", f.Namespace, f.Name);
+                        if (request.RequestUri.OriginalString.Contains(fname))
+                        {
+                            isFunctionCall = true;
+                            break;
+                        }
+                    }
+                    // only, if it is not a function call
+                    if (!isFunctionCall)
+                    {
+                        var response = request.GetOwinContext()?.Response;
+                        response?.OnSendingHeaders(state =>
+                        {
+                            ((IOwinResponse) state).StatusCode = (int) HttpStatusCode.NotFound;
+                        }, response);
+                        // in case you are NOT using Owin, uncomment the following and comment everything above
+                        // HttpContext.Current.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                    }
+                    return _nullEntityTypeSerializer;
+                }
+                return serializer;
+            }
+        }
+
+        public class NullEntityTypeSerializer : ODataEntityTypeSerializer
+        {
+            public NullEntityTypeSerializer(ODataSerializerProvider serializerProvider)
+                : base(serializerProvider)
+            { }
+            public override void WriteObjectInline(object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)
+            {
+                if (graph != null)
+                {
+                    base.WriteObjectInline(graph, expectedType, writer, writeContext);
+                }
+            }
+        }
+
         public class SharedModel
         {
             [Key]
@@ -93,6 +178,11 @@ namespace Swashbuckle.OData.Tests
             public string Variation { get; set; }
         }
 
+        public class CustomApiModel
+        {
+            public int PropertyA { get; set; }
+        }
+
         public class SharedModelsController : ODataController
         {
             [EnableQuery]
@@ -123,6 +213,17 @@ namespace Swashbuckle.OData.Tests
                 };
                 return sharedModels;
             }
+
+            [Route("CustomApiModels")]
+            public List<CustomApiModel> GetCustomApiModels()
+            {
+                var customApiModels = new List<CustomApiModel>
+                {
+                    new CustomApiModel {PropertyA = 1},
+                    new CustomApiModel {PropertyA = 2}
+                };
+                return customApiModels;
+            }
         }
     }
 

```

**Comment by rbeauchamp-lendingtree** [comment]
Hey @kjst  Thanks for posting this issue! Would be helpful for whomever might fix this to have a pull request that includes your unit test.

**Comment by kfstorm** [comment]
Hi @rbeauchamp-lendingtree , check the pull request: https://github.com/rbeauchamp/Swashbuckle.OData/pull/96

**Comment by rbeauchamp-lendingtree** [comment]
Fix is available in release [v2.18.3.](https://www.nuget.org/packages/Swashbuckle.OData/2.18.3)

Thank you for the fix @kfstorm.

**Comment by kfstorm** [comment]
Glad to hear that!
